External Login registration without password

MichaelPetrinolis opened the pull request:
Add a setting to register external users without requesting a password
![image](https://user-images.githubusercontent.com/9623255/56429405-67bf8a80-62cb-11e9-989d-83b3251a3b93.png)
Generate a username instead of parsing claims from external login info
Integrate Recaptcha in External Login Registration
![image](https://user-images.githubusercontent.com/9623255/56429744-b28dd200-62cc-11e9-939e-9930793b2270.png)
Integrate Email Confirmation for users created from External Logins

MichaelPetrinolis commented:
I must check all the scenarios after merging with dev

sebastienros commented:
The PR shows all the changes since you created the PR, can you rebase it on top of dev?

hishamco commented:
@MichaelPetrinolis could you please rebase on dev and fix the conflict?

MichaelPetrinolis commented:
@hishamco too many conflicts. I had removed the base email controller. Will try to reapply the logic on top of the latest version as soon as i can.

sebastienros commented:
I fixed the conflicts, merged dev, tested it.
Then I added the options to not update the Email and Username once externally authenticated.
If the 3 options are enabled then we can skip the external login confirmation and go directly to the consent (if any).

@PinpointTownes I'd like your feedback on this.

MichaelPetrinolis commented:
Rebased on latest dev and refactored the code to be more simple.

Made a QA test and seems to work as expected. The only issue was when trying to access the admin and registered a login from an external provider without access to admin panel, there was a recursive loop. I solved it by checking if the current user is authenticated and reset the returnUrl. If the mapping feature is merged, the admin could define roles for new registrations.

Added a LoginSetting to disable local logins and use the only one defined external provider for local users. I also added an option to provide the default username based on a server script.

sebastienros commented:
Demo on Tuesday?

MichaelPetrinolis commented:
I will try.

sebastienros commented:
Are there anything to add to this PR based on some comments we could have made during the meeting? I don't remember.

MichaelPetrinolis commented:
I can't recall other than [hd](https://developers.google.com/identity/protocols/OpenIDConnect#hd-param) parameter discussion for Google Id. I would like to add a setting to explicitly disable local logins. Now this happens when you configure the 'UseExternalProviderIfOnlyOneDefined' in login settings. A was thinking introducing two settings. One DisableLocalLogins and another one to define the name of external provider to use as default. If disable local logins and no external providers, the disable local login setting will have no effect. If no external provider name is given, a list of external providers will be presented to the user to select.

@PinpointTownes what do you think of this PR ?

sebastienros commented:
@PinpointTownes demo here: https://youtu.be/N50qi1NFZXU?t=1453

MichaelPetrinolis commented:
There is no check in disabling local login. There is no check in deleting the Administrator role neither.
Should we add validations to make sure that there will always be an admin in the system and disable the local login only if there is an admin user with linked external provider?

MichaelPetrinolis commented:
@sebastienros could we merge this PR? There should be no issues...

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:
```diff
@@ -199,107 +293,239 @@ public async Task<IActionResult> ExternalLoginCallback(string returnUrl = null,
+            if (settings.UsersCanRegister == RegistrationSettings.UsersCanRegisterEnum.NoRegistration)
             {
-                _logger.LogInformation("Site does not allow user registration.");
+                _logger.LogWarning("Site does not allow user registration. {externalLoginViewModel}", model);
```
We should not log it this way as it contains private information. Better pick the field that makes sense.

sebastienros replied:
There are other places where the view model is passed in the log btw.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Users/Controllers/ResetPasswordController.cs:
```diff
@@ -20,7 +20,7 @@
 {
     [Feature("OrchardCore.Users.ResetPassword")]
-    public class ResetPasswordController : BaseEmailController
+    public class ResetPasswordController : Controller
```
Thanks

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Users/ViewModels/RegisterExternalLoginViewModel.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            if (string.IsNullOrWhiteSpace(Email) && !NoEmail)
+            {
+                yield return new ValidationResult("Email is required!", new[] { "Email" });
```
Shouldn't it be string-localized? Have you seen this pattern somewhere else that we should fix too?

MichaelPetrinolis replied:
Opened an issue #4619

hishamco commented on src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:
```diff
@@ -282,7 +282,7 @@ public async Task<IActionResult> ExternalLoginCallback(string returnUrl = null,
             if (remoteError != null)
             {
-                _logger.LogError("Error from external provider: {Error}", remoteError);
+                _logger.LogError($"Error from external provider: {remoteError}");
```
Localize the log message

sebastienros replied:
Well this change is wrong, revert it.

MichaelPetrinolis replied:
I merged dev branch, keep target accountcontroller and manually did the changes. But when I commited the merge, the AccountController was not staged and left uncommited.
That happens when instead of sleeping you merge :)

MichaelPetrinolis commented:
@sebastienros  i have the following suggestion. How about introducing the ExternalLoginEvent ( ExternalLoginEventArgs )  event with ExternalLoginEventArgs { Provider, Claims, Handled, UsernameFormat, LocalRoles } that will be triggered on external logins?

I could merge the #3722 (Map external claims) into this one and add a setting to handle the event
by setting up a js script in the settings if we want to allow oob customization. Otherwise, it could be customized from a module by registering an event handler.

hishamco commented on src/OrchardCore.Modules/OrchardCore.OpenId/Views/Access/Authorize.cshtml:
```diff
@@ -1,10 +1,14 @@
 @model AuthorizeViewModel
 
+@{
+    ViewLayout = "Layout__Login";
```
Could we move it to `_ViewStart.cshtml`?

hishamco replied:
It's `Layout` or `ViewLayout`?!!

MichaelPetrinolis replied:
It is a trick to use the layout from front end theme instead of the admin. It is used on all login views

hishamco replied:
I see

hishamco commented on src/OrchardCore.Modules/OrchardCore.Users/Controllers/ControllerExtensions.cs:
```diff
@@ -0,0 +1,108 @@
+
+namespace OrchardCore.Users.Controllers
+{
+    internal static class ControllerExtensions
```
I dislike this one, why we don't use `ISmtpService` directly to send emails, same thing for others

MichaelPetrinolis replied:
There was a base controller that account and registration inherited from that one, only to reuse the email functionality. In order to be reusable and remove the base controller, i created that extension class.

hishamco replied:
It's fine while it's `internal` :)

hishamco commented on src/OrchardCore.Modules/OrchardCore.Users/Controllers/RegistrationController.cs:
```diff
@@ -38,20 +37,16 @@ public class RegistrationController : BaseEmailController
             IHtmlLocalizer<RegistrationController> htmlLocalizer,
-            IStringLocalizer<RegistrationController> stringLocalizer,
-            IEnumerable<IRegistrationFormEvents> registrationEvents) : base(smtpService, displayHelper)
+            IStringLocalizer<RegistrationController> stringLocalizer)
```
Where this one is used?!!

hishamco commented on src/OrchardCore.Modules/OrchardCore.Users/Models/RegistrationSettings.cs:
```diff
@@ -2,8 +2,19 @@ namespace OrchardCore.Users.Models
     public class RegistrationSettings
     {
-        public bool UsersCanRegister { get; set; }
+        public enum UsersCanRegisterEnum
```
Extract to a new file

Can we remove the Enum suffix and use something like `RegistrationType` or something similar

hishamco commented on src/OrchardCore.Modules/OrchardCore.Users/ViewModels/LinkExternalLoginViewModel.cs:
```diff
@@ -0,0 +1,23 @@
+
+        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
+        {
+            var S = validationContext.GetService<IStringLocalizer<LinkExternalLoginViewModel>>();
```
I think using DI is better, so we can use the localizer everywhere in the class

MichaelPetrinolis replied:
I agree, but then we should register the view models with DI and inject them in constructors.

hishamco replied:
> but then we should register the view models with DI 

Why? What I want to inject the `IStringLocalizer` in the `LinkExternalLoginViewModel`

MichaelPetrinolis replied:
How would you inject the IStringLocalizer to the LinkExternalLoginViewModel without registering LinkExternalLoginViewModel in the container and request a resource of type IStringLocalizer in it's constructor ?

hishamco replied:
Ok, just leave my previous comment

hishamco commented on src/OrchardCore.Modules/OrchardCore.Users/Views/Account/Login.cshtml:
```diff
@@ -6,57 +6,60 @@
 @inject SignInManager<IUser> _signInManager
 @inject ISiteService _site
-
+@inject UserManager<IUser> _userManager;
```
Nah, I prefer to use `UserManager` instead

hishamco commented on src/OrchardCore.Modules/OrchardCore.Users/Views/Account/Login.cshtml:
```diff
@@ -6,57 +6,60 @@
     @if (loginProviders.Count > 0)
     {
-        <div class="col-md-6">
+        <div class="col-md-6 @(disableLocalLogin ? "offset-md-3" : "")">
```
String.Empty

hishamco commented on src/OrchardCore.Modules/OrchardCore.Users/Views/Account/Login.cshtml:
```diff
@@ -65,7 +68,11 @@
                             {
-                                <button type="submit" class="btn btn-primary" name="provider" value="@provider.Name" title="Log in using your @provider.DisplayName account">@provider.DisplayName</button>
+                                <button type="submit" class="btn btn-primary @(disableLocalLogin ? "w-100" : "")" name="provider" value="@provider.Name" title="Log in using your @provider.DisplayName account">@provider.DisplayName</button>
+                                @if (disableLocalLogin)
```
Is the <br/> for a reason?!!

hishamco commented on src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml:
```diff
@@ -1,27 +1,159 @@
+    function resetScript(keepText) {
+        var editor = $('#@Html.IdFor(x => x.GenerateUsernameScript)').data('editor');
+        if (!keepText) {
+            editor.doc.setValue(
```
I don't understand what this is used for!!

MichaelPetrinolis replied:
When there is no script saved in the settings, it initializes the default script on load. 
@hishamco thanx for the review.

sebastienros commented:
@MichaelPetrinolis should I merge this?

MichaelPetrinolis commented:
@sebastienros give some more time to integrate  #3722 in this PR and introduce the ExternalLoginEvent. I

MichaelPetrinolis commented:
I Integrated #3722 into this PR. Let me make a synopsis of what this adds to the OrchardCore.Users

BTW the new menu structure is fantastic!

Login Settings
![image](https://user-images.githubusercontent.com/9623255/69484543-f6395300-0e3c-11ea-8edf-87015065eb5a.png)

Use External Provider To login : If only one external provider is defined and this option is checked, the external provider is automatically challenged.
Disable local password login: Prevents users from logging in with their local credentials. Only external providers are displayed in the login screen, unless the above option is set.

Introduced IExternalLoginEventHandler , with UpdateRoles and GeneratePassword methods. Any module can define logic to perform the mapping and generate usernames.

Use a script to set user roles based on external provider claims: If checked, the above handlers are not invoked and instead the script executes.

Introduced the UserLoggedIn Workflow event. The workflow is triggered after the above logic, and provides the username,the roles, the login provider and the external claims as Input. If the login was for local user, the login provider and the external claims are null

Registration Settings
![image](https://user-images.githubusercontent.com/9623255/69485520-5504c980-0e49-11ea-93f3-b66388c35603.png)

Configure users registration: Define if new users are allowed to register.

External Authentication

When a new user is registering, the registered IExternalLoginEventHandlers are invoked in order to provide a username. If more than one handlers are defined, the first one wins.

If you use a script to generate the userName, the above handlers are not triggered.

Do not ask username : if checked, no user name is requested in the registration form
Do not ask email address: if checked, no email address is requested in the registration form
Do not create local password: if checked, no password is requested in the registration form

If all of the above are true, no registration form is displayed to the user.

carlwoodhouse commented:
this looks pretty awesome :)

nuno-tiago-reis commented:
Does IExternalLoginEventHandler (more specifically UpdateRoles(UpdateRolesContext)) allow complex JSON claims or only simple type claims?

nuno-tiago-reis commented:
@MichaelPetrinolis 

Does IExternalLoginEventHandler (more specifically UpdateRoles(UpdateRolesContext)) allow complex JSON claims or only simple type claims?

MichaelPetrinolis commented:
@nuno-tiago-reis Have you tried and faced any issue? The claims are serialized so you can handle them as you wish, either from js script or a custom module that implements the IExternalLoginEventHandler

nuno-tiago-reis commented:
I tried it but I couldn't find the claims that had complex JSON objects or boolean values in the ExternalClaims (for example email_verified and a custom claim that I used to provide some information about the users structure that's necessary for authorization).
Every other claim seems to be there.

MichaelPetrinolis commented:
if you run from orchardcore source, please check the claims you received from your idp.

https://github.com/OrchardCMS/OrchardCore/blob/22e7ba60cab45faf3d2badf5cb1e60675a063d8b/src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs#L311

nuno-tiago-reis commented:
Yeah I just realized I wasn't requesting the email verified claim but the custom claims are being requested.

Unfortunately I'm not running from the source, is there any other way to check what the actual token is?

MichaelPetrinolis commented:
you can use any openid connect client and inspect the claims.

nuno-tiago-reis commented:
Alright, sorry about the mixup, I didn't configured my OIDC client correctly.

Thanks for the help :)

nuno-tiago-reis commented:
@MichaelPetrinolis 

Is there any way to "shortcircuit" the process from the IExternalLoginEventHandler interface? More specifically the UpdateRoles method. Currently if the user has no role that can access the administrator section and what happens is that the login redirection keeps looping and never exists.

MichaelPetrinolis commented:
@nuno-tiago-reis please file a bug. It should redirect the user to the homepage after login, if the user has no permission to access the control panel.
